File: tests/fixtures/ldap_injection/ruby/unsafe_ldap_search.rb
# Unsafe: tainted Rails param interpolated into the LDAP filter passed to
# Net::LDAP#search.  The receiver is constructed via Net::LDAP.new and
# carries TypeKind::LdapClient; type-qualified resolution rewrites
# `ldap.search` → `LdapClient.search`, firing LDAP_INJECTION.
require "net/ldap"

class UsersController
  def lookup(params)
    ldap = Net::LDAP.new(host: "ldap.example.com")
    user = params[:user]
    filter = "(uid=#{user})"
    ldap.search(base: "ou=people,dc=example,dc=com", filter: filter)
  end
end

File: tests/fixtures/ldap_injection/java/UnsafeLdapSearch.java
// Unsafe: attacker-controlled username concatenated into an LDAP filter passed
// to DirContext.search.  The receiver `ctx` carries TypeKind::LdapClient via
// the declared `DirContext` type so type-qualified resolution rewrites the
// callee to `LdapClient.search` and the LDAP_INJECTION sink fires.
import javax.naming.directory.DirContext;
import javax.naming.directory.SearchControls;
import javax.servlet.http.HttpServletRequest;

public class UnsafeLdapSearch {
    private DirContext ctx;

    public Object lookup(HttpServletRequest req) throws Exception {
        String user = req.getParameter("user");
        String filter = "(uid=" + user + ")";
        return ctx.search("ou=people,dc=example,dc=com", filter, new SearchControls());
    }
}
